### PR TITLE
how-to for setting up the lighting device on Nordic nrf52840 dongle

### PR DESCRIPTION
**Pull request by MonicaisHer**
The purpose of this PR is to facilitate easier review and feedback on the content wiki page. The content of this PR will eventually be moved to the wiki page of this repository but will not be merged into the main branch.

### CODE REVIEW

File: Thread-device.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+guide's steps 1 and 2 to install `nRF command-line tools`, `GN meta-build system`, and `nRF Connect SDK`.
+
+Once nRF Desktop is installed, change its permissions, install dependencies, and run it:
+```bash
+chmod +x nrfconnect-4.2.0-x86_64.appimage
+sudo apt install libfuse2
+./nrfconnect-4.2.0-x86_64.appimage
+```
+
+In nRF Connect for Desktop, install the latest `toolchain manager` and `nRF Connect SDK v2.4.2`.
```
Comment (farshidtz) [inline Thread-device.md:29]: These should not be codified. 
```suggestion
In nRF Connect for Desktop, install the latest **Toolchain Manager** and **nRF Connect SDK v2.4.2**. 
```

File: Thread-device.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+Once nRF Desktop is installed, change its permissions, install dependencies, and run it:
+```bash
+chmod +x nrfconnect-4.2.0-x86_64.appimage
+sudo apt install libfuse2
+./nrfconnect-4.2.0-x86_64.appimage
+```
+
+In nRF Connect for Desktop, install the latest `toolchain manager` and `nRF Connect SDK v2.4.2`. 
+After installations, click the down arrow next to the nRF Connect SDK version you installed and select `Open Terminal` or `Open Bash`.
```
Comment (farshidtz) [inline Thread-device.md:30]:
```suggestion
After installations, click the down arrow next to the nRF Connect SDK version you installed and select **Open Terminal** or **Open Bash**.
```

File: Thread-device.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+### 2. Installing and Activating Matter SDK:
```
Comment (farshidtz) [inline Thread-device.md:32]: Section headings do not need colons.
```suggestion
### 2. Installing and Activating Matter SDK
```

File: Thread-device.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+scripts/checkout_submodules.py --shallow --platform linux
+git submodule update --init
+```
+
+Activate Matter build environment:
+```bash
+cd ~/connectedhomeip/
+
+# Skip all non-core Python requirements:
+sed -i '/^-r requirements/ s/./#&/' ./scripts/setup/requirements.txt
```
Comment (farshidtz) [inline Thread-device.md:47]: Missing an explanation on why this is needed here (i.e. the python incompatibility error preventing the build later on).

File: Thread-device.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+Complete the following steps to build and flash the firmware:
+
+### Environment Setup
+- Ubuntu 23.04 OS
+- nRF52840 dongle
+
+### Version tested in this guide
+- Nordic Semiconductor's nRF Connect for Desktop: 4.2.0
+- Nordic Semiconductor's nRF Connect SDK: 2.4.2
+- [Matter](https://github.com/project-chip/connectedhomeip) SDK: 1.1.0.1
```
Comment (farshidtz) [inline Thread-device.md:15]:
```suggestion
- [Matter](https://github.com/project-chip/connectedhomeip SDK): 1.1.0.1
```

File: Thread-device.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+After installations, click the down arrow next to the nRF Connect SDK version you installed and select `Open Terminal` or `Open Bash`.
+
+### 2. Installing and Activating Matter SDK:
+Install the Matter SDK with a shallow clone for the Linux platform:
+```bash
+cd ~
+git clone https://github.com/project-chip/connectedhomeip.git --depth=1 --branch=v1.1.0.1
+cd ~/connectedhomeip/
+scripts/checkout_submodules.py --shallow --platform linux
+git submodule update --init
```
Comment (farshidtz) [inline Thread-device.md:39]: These two commands work against each other. The first command makes a shallow clone of Linux platform projects, the second makes a deep clone of everything.

I'll do some further testing to see if the first command has any extra effect.

File: Thread-device.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+```bash
+cd ~/connectedhomeip/
+
+# Skip all non-core Python requirements:
+sed -i '/^-r requirements/ s/./#&/' ./scripts/setup/requirements.txt
+source scripts/activate.sh
+```
+### 3. Building the Matter Lighting Example Firmware:
+By default, support for DFU (Device Firmware Upgrade) using Matter OTA is enabled.
+To ensure successful firmware building for the dongle, utilize the following command, considering that nRF52840 dongle doesn't have the external flash required for DFU:
```
Comment (farshidtz) [inline Thread-device.md:52]: Hard to follow.
```suggestion
This needs to be disabled for the nRF52840 dongle because it doesn't have the external flash required for DFU:
```

File: Thread-device.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+### 2. Installing and Activating Matter SDK:
+Install the Matter SDK with a shallow clone for the Linux platform:
+```bash
+cd ~
+git clone https://github.com/project-chip/connectedhomeip.git --depth=1 --branch=v1.1.0.1
+cd ~/connectedhomeip/
+scripts/checkout_submodules.py --shallow --platform linux
+git submodule update --init
+```
+
```
Comment (farshidtz) [inline Thread-device.md:41]: If you get `ImportError: libffi.so.7: cannot open shared object file: No such file or directory` during activation, install libffi7 as described [here](https://askubuntu.com/a/1300763/366811).

File: Thread-device.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+Activate Matter build environment:
+```bash
+cd ~/connectedhomeip/
+
+# Skip all non-core Python requirements:
+sed -i '/^-r requirements/ s/./#&/' ./scripts/setup/requirements.txt
+source scripts/activate.sh
```
Comment (farshidtz) [inline Thread-device.md:48]: I am getting the following error:
```
File "/home/farshid/ncs/v2.4.2/connectedhomeip/.environment/pigweed-venv/lib/python3.8/site-packages/pw_rpc/internal/packet_pb2.py", line 33, in <module>
    _descriptor.EnumValueDescriptor(
  File "/home/farshid/ncs/toolchains/1f9b40e71a/usr/local/lib/python3.8/site-packages/google/protobuf/descriptor.py", line 796, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/farshid/ncs/v2.4.2/connectedhomeip/.environment/pigweed-venv/bin/pw", line 8, in <module>
    sys.exit(main())
  File "/home/farshid/ncs/v2.4.2/connectedhomeip/.environment/pigweed-venv/lib/python3.8/site-packages/pw_cli/__main__.py", line 41, in main
    pw_command_plugins.register()
  File "/home/farshid/ncs/v2.4.2/connectedhomeip/.environment/pigweed-venv/lib/python3.8/site-packages/pw_cli/pw_command_plugins.py", line 64, in register
    _register_builtin_plugins(_plugin_registry)
  File "/home/farshid/ncs/v2.4.2/connectedhomeip/.environment/pigweed-venv/lib/python3.8/site-packages/pw_cli/pw_command_plugins.py", line 41, in _register_builtin_plugins
    registry.register_by_name('test', 'pw_unit_test.test_runner', 'main')
  File "/home/farshid/ncs/v2.4.2/connectedhomeip/.environment/pigweed-venv/lib/python3.8/site-packages/pw_cli/plugins.py", line 303, in register_by_name
    Plugin.from_name(name, module_name, member_name, source)
  File "/home/farshid/ncs/v2.4.2/connectedhomeip/.environment/pigweed-venv/lib/python3.8/site-packages/pw_cli/plugins.py", line 104, in from_name
    raise Error(f'Failed to import module "{module_name}"') from err
pw_cli.plugins.Error: Failed to import module "pw_unit_test.test_runner" (TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates)
```
It can be avoided by setting the environment variable suggested in the error message:
`PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python source scripts/activate.sh`

File: Thread-device.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+```
+### 3. Building the Matter Lighting Example Firmware:
+By default, support for DFU (Device Firmware Upgrade) using Matter OTA is enabled.
+To ensure successful firmware building for the dongle, utilize the following command, considering that nRF52840 dongle doesn't have the external flash required for DFU:
+```bash
+cd examples/lighting-app/nrfconnect
+west build -b nrf52840dongle_nrf52840 -- -DCONF_FILE=./prj_no_dfu.conf
+```
+
```
Comment (farshidtz) [inline Thread-device.md:57]: I got an error here suggesting that I should upgrade my environment by running `/home/farshid/ncs/v2.4.2/connectedhomeip/scripts/setup/nrfconnect/update_ncs.py --update`. It resolved the error.

```suggestion

Note that the build may fail, suggesting an upgrade of the NCS environment via a given command. Executing the command upgrades the environment and resolves the error.
```

File: Thread-device.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+Upon successful flashing, logs will display:
+```bash
+Uploading image through SDFU: 100%
+All DFU images have been written to the target device
+Target device closed
+```
+
+### 5. Running the Matter Thread Lighting Device:
+Connect the dongle to a USB port, then use `minicom` to run the lighting app:
```
Comment (farshidtz) [inline Thread-device.md:76]:
```suggestion
Connect the dongle to a USB port, then use `minicom` to open a serial consol and start the lighting app:
```

**Review by farshidtz — CHANGES_REQUESTED** [review]
Looks good overall. Some editorial issues and missing pieces to make it work on the given environment.

File: Thread-device.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (MonicaisHer) [reply Thread-device.md]: Updated to: "In nRF Connect for Desktop, first, install Toolchain Manager. Then, open Toolchain Manager to install nRF Connect SDK v2.4.2."

File: Thread-device.md (hunk 6 above)
Comment (MonicaisHer) [reply Thread-device.md:40]: Does `git submodule update --init` perform a deep clone?  It only initializes and updates submodules. 

 a deep clone of submodules will be `git submodule update --init --recursive`

Comment (farshidtz) [reply Thread-device.md:40]: By setting `--recursive`, it will update also the nested submodules; it isn't about the history depth. 
The shallow/deep history isn't affecting functionality for the build, but just the size of the history.

Anyway, the second command clones all submodules and the first one just the ones for Linux platform. What we want here is indeed all modules, or only those relevant to this build.

**Review by farshidtz — APPROVED** [review]
Thanks. It is ready to move over to the wiki. 

The one remaining comment can be addressed later: https://github.com/canonical/openthread-border-router-snap/pull/25#discussion_r1348780133
